pkg/power/power.go

```diff
@@ -0,0 +1,312 @@
+
+	if power.Target.IsUp {
+		if power.State.IsDown {
+			_, err = cl.SetNodePowerState(p.identity, true)
+			return errors.Wrap(err, "failed to set state to up")
+		}
+
+		return nil
+	}
+
+	// now the target must be down.
+	// we need to shutdown
+
+	if power.State.IsUp {
+		_, err = cl.SetNodePowerState(p.identity, false)
+		return errors.Wrap(err, "failed to set state to up")
+	}
+
+	// otherwise node need to get back to sleep.
+	if err := p.shutdown(); err != nil {
+		return errors.Wrap(err, "failed to issue shutdown")
+	}
+
+	return nil
+}
+
+func (p *PowerServer) powerUp(node *substrate.Node, reason string) error {
+	log.Info().Uint32("node", uint32(node.ID)).Str("reason", reason).Msg("powering on node")
+
+	mac := ""
+	for _, inf := range node.Interfaces {
+		if inf.Name == "zos" {
+			mac = inf.Mac
+			break
+		}
+	}
+	if mac == "" {
+		return fmt.Errorf("can't find mac address of node '%d'", node.ID)
+	}
+
+	return exec.Command("ether-wake", "-i", "zos", mac).Run()
+
+}
+
+func (p *PowerServer) shutdown() error {
+	if !p.enabled {
+		log.Info().Msg("ignoring shutdown because power managed is not enabled")
+		return nil
+	}
+
+	log.Info().Msg("shutting down node because of chain")
+	if _, err := p.ut.SendNow(); err != nil {
+		log.Error().Err(err).Msg("failed to send uptime before shutting down")
+	}
+
+	// is down!
+	init := zinit.Default()
+	err := init.Shutdown()
+
+	if errors.Is(err, zinit.ErrNotSupported) {
+		log.Info().Msg("node does not support shutdown. rebooting to update")
+		return init.Reboot()
+	}
+
+	return err
+}
+
+func (p *PowerServer) event(event *pkg.PowerTargetChangeEvent) error {
+	if event.FarmID != p.farm {
+		return nil
+	}
+
+	log.Debug().
+		Uint32("farm", uint32(p.farm)).
+		Uint32("node", p.node).
+		Msg("received power event for farm")
+
+	cl, err := p.sub.Substrate()
+	if err != nil {
+		return err
+	}
+	node, err := cl.GetNode(event.NodeID)
+	if err != nil {
+		return err
+	}
+
+	if event.NodeID == p.node && event.Target.IsDown {
+		// we need to shutdown!
+		if _, err := cl.SetNodePowerState(p.identity, false); err != nil {
+			return errors.Wrap(err, "failed to set node power state to down")
+		}
+
+		return p.shutdown()
+	} else if event.Target.IsDown {
+		return nil
+	}
+
+	if event.Target.IsUp {
+		log.Info().Uint32("target", event.NodeID).Msg("received an event to power up")
+		return p.powerUp(node, "target is up")
+	}
+
+	return nil
+}
+
+func (p *PowerServer) recv(ctx context.Context) error {
+	log.Info().Msg("listening for power events")
+	stream, err := p.consumer.PowerTargetChange(ctx)
+	if err != nil {
+		return errors.Wrapf(errConnectionError, "failed to connect to zbus events: %s", err)
+	}
+
+	for event := range stream {
+		if err := p.event(&event); err != nil {
+			log.Error().Err(err).Msg("failed to process power event")
+		}
+	}
+
+	return nil
+}
+
+// start processing time events.
+func (p *PowerServer) events(ctx context.Context) error {
```

Comment:
would opening and closing the substrate connection here help a bit??

Reply:
i'm suggesting this because if there were many power management events, with each event being processed a substrate connection is opened, which may be costly???

Reply:
while the node receive all power events, not all events reaches here, only events intended to the node farm are actually propagated to this code